### ISSUE TITLE
C++ : extern template on function fails at parsing

### ISSUE THREAD

**Issue by romintomasetti** [swig/swig#2131]
Hi there !

# Issue

I think that SWIG cannot manage `extern template` with functions...

According to http://www.swig.org/Doc4.0/CPlusPlus11.html#CPlusPlus11_extern_template, it should work for classes at least.

# Minimal non-working example

The following example works (on Ubuntu 20). But if you comment the `#ifdef SWIG / #endif` block in `library.hpp`, `SWIG` will fail parsing with
```text
library.hpp:11: Error: Syntax error in input(1).
```

There are 4 files, that I put hereinbelow.

`compile.sh` (run it with `bash compile.sh`)
```bash
# Fail if a command fails
set -e

# Remove artifacts
rm *.o *.cxx *.py *.so || true

# Compile C++ files
g++ -std=c++11 -fPIC -c main.cpp library.cpp

# Compile a C++ executable and run it
g++ main.o library.o -o a.out
./a.out

# Parse with SWIG, will fail if in 'library.hpp' you comment the '#ifndef SWIG / #endif' lines
swig -c++ -python library.i

# Compile SWIG-generated C++ files
g++ -std=c++11 -fPIC -c library_wrap.cxx -I/usr/include/python3.8 -o library_wrap.o

# Create Python library
g++ -shared library.o library_wrap.o -o _library.so -L/usr/lib -lpython3.8

# Run the example
python -c "import library;library.test()"
```

`main.cpp`
```c++
#include "library.hpp"

int main()
{
    // BigFunction symbol is in 'library.o' thanks to explicit template 
    BigFunction<int>();
    return 0;
}
```

`library.hpp`
```c++
#ifndef LIBRARY_HPP_
#define LIBRARY_HPP_

// BigFunction is only declared, the code is defined in 'library.cpp'
template<typename T>
void BigFunction();

// If you comment the '#ifndef SWIG / #endif' block, SWIG will fail
#ifndef SWIG
// extern template is valid C++ code
extern template void BigFunction<int>();
#endif

#endif // LIBRARY_HPP_
```

`library.cpp`
```c++
#include <iostream>

#include "library.hpp"

// BigFunction definition
template<typename T>
void BigFunction()
{
    std::cout << "Hello, I am a template with " << typeid(T).name() << " from " << __PRETTY_FUNCTION__ << std::endl;
}

// Explicit template instantiation of BigFunction for 'int'
template void BigFunction<int>();
```

`library.i`
```swig
%module library

%{
    #include "library.hpp"
%}

%include "library.hpp"

%inline %{
    void test()
    {
        BigFunction<int>();
    }
%}
```

**Comment by ojwb** [comment]
Looking at `Source/CParse/parser.y` the grammar only has a rule for `extern template` for the class case.

Seems to be a missing C++11 feature.

**Comment by wsfulton** [comment]
Closed by #2157
